ENIGMAsystem/SHELL/Universal_System/Extensions/ProcInfo/procinfo.h

```diff
@@ -0,0 +1,111 @@
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ SOFTWARE.
+ 
+*/
+
+#include <string>
+#include <vector>
+
+// individual platforms need their platform-specific
+// window types casted to an unsigned long long type
+#define window_t unsigned long long
+
+// we don't really care about the process swapper so
+// if (pid == 0) an error occured with some function
+#define process_t unsigned long
+
+// an unsigned long long window identifier placed in
+// quotes for quick concatenation in shell scripting
+#define wid_t std::string
+
+bool string_has_whitespace(std::string str);
+std::string string_replace_all(std::string str, std::string substr, std::string nstr);
+std::vector<std::string> string_split(std::string str, char delimiter);
+
+namespace procinfo {
+
+process_t process_execute(std::string command);
+void process_execute_async(std::string command);
+process_t process_previous();
+std::string process_evaluate();
+void process_clear_pid();
+void process_clear_out();
+process_t pid_from_self();
+process_t ppid_from_self();
+std::string path_from_pid(process_t pid);
+std::string dir_from_pid(process_t pid);
+std::string name_from_pid(process_t pid);
+std::string cmd_from_pid(process_t pid);
+std::string env_from_pid(process_t pid);
+bool pid_exists(process_t pid);
+bool wid_exists(wid_t wid);
+bool pid_kill(process_t pid);
+window_t window_from_wid(wid_t wid);
+wid_t wid_from_window(window_t window);
+process_t pid_from_wid(wid_t wid);
+std::string pids_enum(bool trim_dir, bool trim_empty);
+process_t ppid_from_pid(process_t pid);
+std::string pids_from_ppid(process_t ppid);
+std::string wids_from_pid(process_t pid);
+wid_t wid_from_top();
+process_t pid_from_top();
+void wid_to_top(wid_t wid);
+void wid_set_pwid(wid_t wid, wid_t pwid);
+std::string echo(std::string expression);
+
+} // namespace procinfo
+
+namespace enigma_user {
+
+using procinfo::process_execute;
+using procinfo::process_execute_async;
+using procinfo::process_previous;
+using procinfo::process_evaluate;
+using procinfo::process_clear_pid;
+using procinfo::process_clear_out;
+using procinfo::pid_from_self;
+using procinfo::ppid_from_self;
+using procinfo::path_from_pid;
+using procinfo::dir_from_pid;
+using procinfo::name_from_pid;
+using procinfo::cmd_from_pid;
+using procinfo::env_from_pid;
+using procinfo::pid_exists;
+using procinfo::wid_exists;
+using procinfo::pid_kill;
+using procinfo::window_from_wid;
+using procinfo::wid_from_window;
+using procinfo::pid_from_wid;
+using procinfo::pids_enum;
+using procinfo::ppid_from_pid;
+using procinfo::pids_from_ppid;
+using procinfo::wids_from_pid;
+using procinfo::wid_from_top;
+using procinfo::pid_from_top;
+using procinfo::wid_to_top;
+using procinfo::wid_set_pwid;
+using procinfo::echo;

```

Comment:
Who told you to keep writing these long using chains?

Reply:
lol this was done by accident because it was copied over from my own repository, this update. Taking care of it as we speak.